sceneorg/views.py

```diff
@@ -163,10 +172,33 @@ def compofiles(request):
             )
         WHERE
             sceneorg_directory.is_deleted = 'f'
+            {exclude_by_id_condition}
         GROUP BY
             sceneorg_directory.id, sceneorg_directory.path
+        {exclude_fully_matched_condition}
         ORDER BY sceneorg_directory.path
-    """)
+    """
+    # Prepare parameters and conditions
+    conditions = {}
+
+    # Handle excluded IDs
+    if excluded_ids:
+        conditions['exclude_by_id_condition'] = 'AND sceneorg_directory.id NOT IN {}'.format(tuple(excluded_ids))
```

Comment:
We should use [the `.raw` method's parameter-passing mechanism](https://docs.djangoproject.com/en/5.2/topics/db/sql/#passing-parameters-into-raw) here, rather than inserting them directly into the SQL. While the risk of SQL injection isn't an issue here (since there is no untrusted user input) it's worth keeping to good habits - and this version fails when there is exactly one excluded ID, leading to the tuple being written as `(12345, )` (which is invalid SQL).

Reply:
Ah, I see the problem - passing a tuple as a parameter doesn't work in version 3 of the postgres driver :-/ https://code.djangoproject.com/ticket/34434

Will go back to a solution that's closer to yours then...